### DISCUSSION TITLE
I can't connect on TCP

### DISCUSSION THREAD

**Discussion by sillo1991** [tr3v3r/react-native-esc-pos-printer#171] [Q&A]
I'm trying to write a simple script to print something but I can't connect to the printer.
Can anyone please help me?

I get the error:  PrinterError: Failed to communicate with the devices within the specified time.

this is my code:

import React, {useState, useMemo} from 'react';
import {Button, Alert} from 'react-native';
import {Printer, PrinterConstants} from 'react-native-esc-pos-printer';

const PrintKitchenOrderButton = () => {
  const [isLoading, setIsLoading] = useState(false);

  const printerInstance = useMemo(
    () =>
      new Printer({
        target: 'TCP:192.168.28.100',
        deviceName: 'TM-T20III',
        port: 8008,
      }),
    [],
  );

  // Function to print
  const printReceipt = async () => {
    try {
      setIsLoading(true);
      const connection = await printerInstance.connect();
      console.log('connection', connection);

      await printerInstance.addTextAlign(PrinterConstants.ALIGN_CENTER);
      await printerInstance.addText('RECEIPT');
      await printerInstance.addFeedLine();
      await printerInstance.addTextAlign(PrinterConstants.ALIGN_LEFT);
      await printerInstance.addText('Item 1');
      await printerInstance.addTextAlign(PrinterConstants.ALIGN_RIGHT);
      await printerInstance.addText('$10.00');
      await printerInstance.addFeedLine();
      await printerInstance.addTextAlign(PrinterConstants.ALIGN_LEFT);
      await printerInstance.addText('Item 2');
      await printerInstance.addTextAlign(PrinterConstants.ALIGN_RIGHT);
      await printerInstance.addText('$20.00');
      await printerInstance.addFeedLine();
      await printerInstance.addCut();

      const result = await printerInstance.sendData();
      await printerInstance.disconnect();
      return result;

      if (res) {
        console.log('Printing completed:', res);
        Alert.alert('Success', 'Printing completed successfully');
      }
    } catch (error) {
      console.error('Error during printing:', error);
      Alert.alert('Error', `Failed to print: ${error.message}`);
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <Button
      title={isLoading ? 'Printing...' : 'Print Kitchen Order'}
      onPress={printReceipt}
      disabled={isLoading}
    />
  );
};

export default PrintKitchenOrderButton;

Thanks a lot in advance

**Comment by sillo1991** [comment]
I have the printer connected to the lan and I'm in the same network, is this the issue maybe?

**Reply by manage-some** [reply]
You will need epson netconfig to configure the printer on you wifi first. 
You can take guide on youtube

**Reply by sillo1991** [reply]
Thanks for the reply, I've already set the static ip address and I was able to print with the epson js sdk... just now I want to make a mobile app and react-native-esc-pos-printer would be great so I don't have to make one code for ios and one for android

**Reply by manage-some** [reply]
download this official app of epson from playstore/appstore

TM Utility

Try to discover and test print with from it. If it works with the official app then it should work with this sdk as well. Otherwise your configuration is not right

**Reply by sillo1991** [reply]
I can print with the epson softwares the test print, but it doesn't work with react-native-esc-pos-printer unfortunately. the error is always  Failed to communicate with the devices within the specified time

**Reply by tr3v3r** [reply]
Have you tried to discover first

**Reply by sillo1991** [reply]
I didn't, I'm going to try soon and let you know shortly

**Reply by sillo1991** [reply]
I've done it but I get Permission error.

I also tried to put all the permissions in android\app\src\main\AndroidManifest.xml
at the moment I put these:

    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
    <uses-permission android:name="android.permission.BLUETOOTH"/>
    <uses-permission android:name="android.permission.BLUETOOTH_ADMIN"/>
    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>

**Comment by tr3v3r** [comment]
Resolved. The issue was in the running not on a real device.
